### ISSUE TITLE
[openssl] Build error with arm64-linux triplet

### ISSUE THREAD

**Issue by mmerah** [microsoft/vcpkg#29674]
**Host Environment**

- OS: Ubuntu Linux 22.04 x64 (running a debian image under docker version 20.10.12)
- Compiler: 
    - g++ (Debian 12.2.0-14) 12.2.0
    - aarch64-linux-gnu-gcc (Debian 12.2.0-9) 12.2.0$
- vcpkg-tool version: 2023-01-24-8a88d63f241d391772fbde69af9cab96c3c64c75
- vcpkg-scripts version: 93895b28e 2023-02-14 (19 hours ago)

**To Reproduce**

```
ARG CROSS_SDK_BASE_TAG=3-bookworm
ARG BASE_VERSION=3-bookworm
ARG IMAGE_ARCH=arm64

FROM torizon/debian-cross-toolchain-${IMAGE_ARCH}:${CROSS_SDK_BASE_TAG} As Build

ARG IMAGE_ARCH
ARG COMPILER_ARCH
ENV IMAGE_ARCH ${IMAGE_ARCH}

RUN apt-get -q -y update && \
    apt-get -q -y install \
    clang-tidy \
    cmake \
    cppcheck \
    curl \
    g++ \
    ninja-build \
    pkg-config \
    python3 \
    unzip \
    zip \
    && \
    apt-get clean && apt-get autoremove && \
    rm -rf /var/lib/apt/lists/*

RUN git clone https://github.com/microsoft/vcpkg
RUN cd vcpkg && \
    ./bootstrap-vcpkg.sh && \
    cd / && \
    chmod -R 777 vcpkg
ENV VCPKG_ROOT="/vcpkg"

RUN cd vcpkg && \
    ./vcpkg install openssl:arm64-linux
```

+ `docker build -t openssl-vcpkg-crosscompile:latest .`

**Failure logs**

```
Installing 3/3 openssl:arm64-linux...
Building openssl[core]:arm64-linux...
warning: -- Using community triplet arm64-linux. This triplet configuration is not guaranteed to succeed.
-- [COMMUNITY] Loading triplet configuration from: /vcpkg/triplets/community/arm64-linux.cmake
CMake Warning at ports/openssl/portfile.cmake:8 (message):
  openssl currently requires the following library from the system package
  manager:

      linux-headers

  It can be installed on alpine systems via apk add linux-headers.
Call Stack (most recent call first):
  scripts/ports.cmake:147 (include)


-- Downloading https://github.com/openssl/openssl/archive/openssl-3.0.8.tar.gz -> openssl-openssl-openssl-3.0.8.tar.gz...
-- Extracting source /vcpkg/downloads/openssl-openssl-openssl-3.0.8.tar.gz
-- Applying patch windows/install-layout.patch
-- Applying patch windows/install-pdbs.patch
-- Applying patch windows/umul128-arm64.patch
-- Using source at /vcpkg/buildtrees/openssl/src/nssl-3.0.8-68504cf2ba.clean
-- Found external ninja('1.11.1').
-- Configuring arm64-linux
-- Building arm64-linux-dbg
CMake Error at scripts/cmake/vcpkg_execute_build_process.cmake:134 (message):
    Command failed: /usr/bin/cmake --build . --config Debug --target install -- -v -j5
    Working Directory: /vcpkg/buildtrees/openssl/arm64-linux-dbg
    See logs for more information:
      /vcpkg/buildtrees/openssl/install-arm64-linux-dbg-out.log

Call Stack (most recent call first):
  installed/x64-linux/share/vcpkg-cmake/vcpkg_cmake_build.cmake:74 (vcpkg_execute_build_process)
  installed/x64-linux/share/vcpkg-cmake/vcpkg_cmake_install.cmake:16 (vcpkg_cmake_build)
  ports/openssl/unix/portfile.cmake:21 (vcpkg_cmake_install)
  ports/openssl/portfile.cmake:34 (include)
  scripts/ports.cmake:147 (include)


error: building openssl:arm64-linux failed with: BUILD_FAILED
error: Please ensure you're using the latest port files with `git pull` and `vcpkg update`.
Then check for known issues at:
    https://github.com/microsoft/vcpkg/issues?q=is%3Aissue+is%3Aopen+in%3Atitle+openssl
You can submit a new issue at:
    https://github.com/microsoft/vcpkg/issues/new?template=report-package-build-failure.md&title=[openssl]+Build+error
Include '[openssl] Build error' in your bug report title, the following version information in your bug description, and attach any relevant failure logs from above.
    vcpkg-tool version: 2023-01-24-8a88d63f241d391772fbde69af9cab96c3c64c75
    vcpkg-scripts version: 93895b28e 2023-02-14 (19 hours ago)


Please use the prefilled template from /vcpkg/installed/vcpkg/issue_body.md when reporting your issue.
```

install-arm64-linux-dbg-out.log :
```
[1/3] cd /vcpkg/buildtrees/openssl/arm64-linux-dbg/nssl-3.0.8-68504cf2ba.clean && export PATH=/usr/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/bin:/usr/bin && export CC=/usr/bin/aarch64-linux-gnu-gcc && export AR=/usr/bin/aarch64-linux-gnu-ar && export LD=/usr/bin/aarch64-linux-gnu-ld && export RANLIB=/usr/bin/aarch64-linux-gnu-ranlib && export MAKE=/usr/bin/make && export MAKEDEPPROG=/usr/bin/aarch64-linux-gnu-gcc && export WINDRES= && /usr/bin/perl Configure no-shared no-module enable-static-engine no-zlib no-ssl2 no-idea no-cast no-seed no-md2 no-tests linux-generic32 --prefix=/vcpkg/packages/openssl_arm64-linux/debug --libdir=/vcpkg/packages/openssl_arm64-linux/debug/lib --openssldir=/etc/ssl "-fPIC -g" && /usr/bin/cmake -DDIR=/vcpkg/buildtrees/openssl/arm64-linux-dbg/nssl-3.0.8-68504cf2ba.clean -P /vcpkg/ports/openssl/unix/remove-deps.cmake
***** Deprecated options: no-ssl2
Configuring OpenSSL version 3.0.8 for target linux-generic32
Using os-specific seed configuration
Created configdata.pm
Running configdata.pm
Created Makefile.in
Created Makefile
Created include/openssl/configuration.h

**********************************************************************
***                                                                ***
***   OpenSSL has been successfully configured                     ***
***                                                                ***
***   If you encounter a problem while building, please open an    ***
***   issue on GitHub <https://github.com/openssl/openssl/issues>  ***
***   and include the output from the following command:           ***
***                                                                ***
***       perl configdata.pm --dump                                ***
***                                                                ***
***   (If you are new to OpenSSL, you might want to consult the    ***
***   'Troubleshooting' section in the INSTALL.md file first)      ***
***                                                                ***
**********************************************************************
removing deps from /vcpkg/buildtrees/openssl/arm64-linux-dbg/nssl-3.0.8-68504cf2ba.clean/demos/bio/Makefile
removing deps from /vcpkg/buildtrees/openssl/arm64-linux-dbg/nssl-3.0.8-68504cf2ba.clean/demos/cipher/Makefile
removing deps from /vcpkg/buildtrees/openssl/arm64-linux-dbg/nssl-3.0.8-68504cf2ba.clean/demos/digest/Makefile
removing deps from /vcpkg/buildtrees/openssl/arm64-linux-dbg/nssl-3.0.8-68504cf2ba.clean/demos/encode/Makefile
removing deps from /vcpkg/buildtrees/openssl/arm64-linux-dbg/nssl-3.0.8-68504cf2ba.clean/demos/kdf/Makefile
removing deps from /vcpkg/buildtrees/openssl/arm64-linux-dbg/nssl-3.0.8-68504cf2ba.clean/demos/mac/Makefile
removing deps from /vcpkg/buildtrees/openssl/arm64-linux-dbg/nssl-3.0.8-68504cf2ba.clean/demos/pkey/Makefile
removing deps from /vcpkg/buildtrees/openssl/arm64-linux-dbg/nssl-3.0.8-68504cf2ba.clean/demos/signature/Makefile
removing deps from /vcpkg/buildtrees/openssl/arm64-linux-dbg/nssl-3.0.8-68504cf2ba.clean/doc/life-cycles/Makefile
[2/3] cd /vcpkg/buildtrees/openssl/arm64-linux-dbg/nssl-3.0.8-68504cf2ba.clean && export PATH=/usr/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/bin:/usr/bin && /usr/bin/cmake -E touch /vcpkg/buildtrees/openssl/arm64-linux-dbg/nssl-3.0.8-68504cf2ba.clean/krb5.h && /usr/bin/make -j 5 build_libs
FAILED: CMakeFiles/build_libs nssl-3.0.8-68504cf2ba.clean/libssl.a nssl-3.0.8-68504cf2ba.clean/libcrypto.a /vcpkg/buildtrees/openssl/arm64-linux-dbg/CMakeFiles/build_libs /vcpkg/buildtrees/openssl/arm64-linux-dbg/nssl-3.0.8-68504cf2ba.clean/libssl.a /vcpkg/buildtrees/openssl/arm64-linux-dbg/nssl-3.0.8-68504cf2ba.clean/libcrypto.a 
cd /vcpkg/buildtrees/openssl/arm64-linux-dbg/nssl-3.0.8-68504cf2ba.clean && export PATH=/usr/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/bin:/usr/bin && /usr/bin/cmake -E touch /vcpkg/buildtrees/openssl/arm64-linux-dbg/nssl-3.0.8-68504cf2ba.clean/krb5.h && /usr/bin/make -j 5 build_libs
/usr/bin/perl "-I." -Mconfigdata "util/dofile.pl" "-oMakefile" include/crypto/bn_conf.h.in > include/crypto/bn_conf.h
...
/usr/bin/perl "-I." -Mconfigdata "util/dofile.pl" "-oMakefile" include/openssl/x509v3.h.in > include/openssl/x509v3.h
/usr/bin/make depend && /usr/bin/make _build_libs
make[1]: Entering directory '/vcpkg/buildtrees/openssl/arm64-linux-dbg/nssl-3.0.8-68504cf2ba.clean'
make[1]: Leaving directory '/vcpkg/buildtrees/openssl/arm64-linux-dbg/nssl-3.0.8-68504cf2ba.clean'
make[1]: Entering directory '/vcpkg/buildtrees/openssl/arm64-linux-dbg/nssl-3.0.8-68504cf2ba.clean'
aarch64-linux-gnu-/usr/bin/aarch64-linux-gnu-gcc  -I. -Iinclude -Iapps/include  -fPIC -pthread -Wall -O3 -fPIC -g -DOPENSSL_USE_NODELETE -DOPENSSL_PIC -DOPENSSLDIR="\"/etc/ssl\"" -DENGINESDIR="\"/vcpkg/packages/openssl_arm64-linux/debug/lib/engines-3\"" -DMODULESDIR="\"/vcpkg/packages/openssl_arm64-linux/debug/lib/ossl-modules\"" -DOPENSSL_BUILDING_OPENSSL -DNDEBUG   -c -o apps/lib/libapps-lib-app_libctx.o apps/lib/app_libctx.c
aarch64-linux-gnu-/usr/bin/aarch64-linux-gnu-gcc  -I. -Iinclude -Iapps/include  -fPIC -pthread -Wall -O3 -fPIC -g -DOPENSSL_USE_NODELETE -DOPENSSL_PIC -DOPENSSLDIR="\"/etc/ssl\"" -DENGINESDIR="\"/vcpkg/packages/openssl_arm64-linux/debug/lib/engines-3\"" -DMODULESDIR="\"/vcpkg/packages/openssl_arm64-linux/debug/lib/ossl-modules\"" -DOPENSSL_BUILDING_OPENSSL -DNDEBUG   -c -o apps/lib/libapps-lib-app_params.o apps/lib/app_params.c
aarch64-linux-gnu-/usr/bin/aarch64-linux-gnu-gcc  -I. -Iinclude -Iapps/include  -fPIC -pthread -Wall -O3 -fPIC -g -DOPENSSL_USE_NODELETE -DOPENSSL_PIC -DOPENSSLDIR="\"/etc/ssl\"" -DENGINESDIR="\"/vcpkg/packages/openssl_arm64-linux/debug/lib/engines-3\"" -DMODULESDIR="\"/vcpkg/packages/openssl_arm64-linux/debug/lib/ossl-modules\"" -DOPENSSL_BUILDING_OPENSSL -DNDEBUG   -c -o apps/lib/libapps-lib-app_provider.o apps/lib/app_provider.c
aarch64-linux-gnu-/usr/bin/aarch64-linux-gnu-gcc  -I. -Iinclude -Iapps/include  -fPIC -pthread -Wall -O3 -fPIC -g -DOPENSSL_USE_NODELETE -DOPENSSL_PIC -DOPENSSLDIR="\"/etc/ssl\"" -DENGINESDIR="\"/vcpkg/packages/openssl_arm64-linux/debug/lib/engines-3\"" -DMODULESDIR="\"/vcpkg/packages/openssl_arm64-linux/debug/lib/ossl-modules\"" -DOPENSSL_BUILDING_OPENSSL -DNDEBUG   -c -o apps/lib/libapps-lib-app_rand.o apps/lib/app_rand.c
/bin/sh: 1: aarch64-linux-gnu-/usr/bin/aarch64-linux-gnu-gcc: not found
make[1]: *** [Makefile:2378: apps/lib/libapps-lib-app_libctx.o] Error 127
make[1]: *** Waiting for unfinished jobs....
/bin/sh: 1: aarch64-linux-gnu-/usr/bin/aarch64-linux-gnu-gcc: not found
/bin/sh: 1: aarch64-linux-gnu-/usr/bin/aarch64-linux-gnu-gcc: not found
make[1]: *** [Makefile:2382: apps/lib/libapps-lib-app_provider.o] Error 127
make[1]: *** [Makefile:2384: apps/lib/libapps-lib-app_rand.o] Error 127
/bin/sh: 1: aarch64-linux-gnu-/usr/bin/aarch64-linux-gnu-gcc: not found
make[1]: *** [Makefile:2380: apps/lib/libapps-lib-app_params.o] Error 127
make[1]: Leaving directory '/vcpkg/buildtrees/openssl/arm64-linux-dbg/nssl-3.0.8-68504cf2ba.clean'
make: *** [Makefile:1395: build_libs] Error 2
ninja: build stopped: subcommand failed.
```

config-arm64-linux-out.log:
```
[2/2] "/usr/bin/cmake" -E chdir "../../arm64-linux-dbg" "/usr/bin/cmake" "/vcpkg/ports/openssl/unix" "-G" "Ninja" "-DCMAKE_BUILD_TYPE=Debug" "-DCMAKE_INSTALL_PREFIX=/vcpkg/packages/openssl_arm64-linux/debug" "-DFETCHCONTENT_FULLY_DISCONNECTED=ON" "-DSOURCE_PATH=/vcpkg/buildtrees/openssl/src/nssl-3.0.8-68504cf2ba.clean" "-DPERL=/usr/bin/perl" "-DMAKE=/usr/bin/make" "-DVCPKG_CONCURRENCY=5" "-DCMAKE_MAKE_PROGRAM=/usr/bin/ninja" "-DCMAKE_SYSTEM_NAME=Linux" "-DBUILD_SHARED_LIBS=OFF" "-DVCPKG_CHAINLOAD_TOOLCHAIN_FILE=/vcpkg/scripts/toolchains/linux.cmake" "-DVCPKG_TARGET_TRIPLET=arm64-linux" "-DVCPKG_SET_CHARSET_FLAG=ON" "-DVCPKG_PLATFORM_TOOLSET=external" "-DCMAKE_EXPORT_NO_PACKAGE_REGISTRY=ON" "-DCMAKE_FIND_PACKAGE_NO_PACKAGE_REGISTRY=ON" "-DCMAKE_FIND_PACKAGE_NO_SYSTEM_PACKAGE_REGISTRY=ON" "-DCMAKE_INSTALL_SYSTEM_RUNTIME_LIBS_SKIP=TRUE" "-DCMAKE_VERBOSE_MAKEFILE=ON" "-DVCPKG_APPLOCAL_DEPS=OFF" "-DCMAKE_TOOLCHAIN_FILE=/vcpkg/scripts/buildsystems/vcpkg.cmake" "-DCMAKE_ERROR_ON_ABSOLUTE_INSTALL_DESTINATION=ON" "-DVCPKG_CXX_FLAGS=" "-DVCPKG_CXX_FLAGS_RELEASE=" "-DVCPKG_CXX_FLAGS_DEBUG=" "-DVCPKG_C_FLAGS=" "-DVCPKG_C_FLAGS_RELEASE=" "-DVCPKG_C_FLAGS_DEBUG=" "-DVCPKG_CRT_LINKAGE=dynamic" "-DVCPKG_LINKER_FLAGS=" "-DVCPKG_LINKER_FLAGS_RELEASE=" "-DVCPKG_LINKER_FLAGS_DEBUG=" "-DVCPKG_TARGET_ARCHITECTURE=arm64" "-DCMAKE_INSTALL_LIBDIR:STRING=lib" "-DCMAKE_INSTALL_BINDIR:STRING=bin" "-D_VCPKG_ROOT_DIR=/vcpkg" "-D_VCPKG_INSTALLED_DIR=/vcpkg/installed" "-DVCPKG_MANIFEST_INSTALL=OFF"
-- Cross compiling arm64 on host x86_64, use cross compiler: aarch64-linux-gnu-g++/aarch64-linux-gnu-gcc
-- The C compiler identification is GNU 12.2.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/aarch64-linux-gnu-gcc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
CMAKE_C_COMPILER=/usr/bin/aarch64-linux-gnu-gcc
COMPILER_ROOT=/usr/bin
CMAKE_SYSROOT=
CMAKE_OSX_SYSROOT=
CMAKE_OSX_DEPLOYMENT_TARGET=
CMAKE_C_FLAGS=-fPIC
CMAKE_C_FLAGS_RELEASE=-O3 -DNDEBUG
CMAKE_C_FLAGS_DEBUG=-g
CMAKE_INCLUDE_SYSTEM_FLAG_C=-isystem 
CMAKE_C_OSX_DEPLOYMENT_TARGET_FLAG=
ENV{ANDROID_DEV}=/usr
-- Configuring done
-- Generating done
CMake Warning:
  Manually-specified variables were not used by the project:

    CMAKE_INSTALL_BINDIR
    CMAKE_INSTALL_LIBDIR
    FETCHCONTENT_FULLY_DISCONNECTED
    VCPKG_PLATFORM_TOOLSET
    VCPKG_SET_CHARSET_FLAG
    _VCPKG_ROOT_DIR


-- Build files have been written to: /vcpkg/buildtrees/openssl/arm64-linux-dbg
```

**Additional context**

For an unknown reason, it seems that `gcc` gets replaced with `/usr/bin/aarch64-linux-gnu-gcc`, which results in this error: `/bin/sh: 1: aarch64-linux-gnu-/usr/bin/aarch64-linux-gnu-gcc: not found`

Edit: In the port Unix/CMakeLists.txt I added `"--cross-compile-prefix=\"\""` to the flags for the `Perl configure` custom command and now it works. I think this is a platform problem because I didn't have that problem with an ubuntu based docker image. Feel free to close if you agree.

**Comment by dg0yt** [comment]
Please check if this is still an issue. https://github.com/microsoft/vcpkg/pull/27261 was merged two days ago. It significantly changed the build for non-windows targets, and it has ` --cross-compile-prefix=`.
(arm64-linux builds for me on x64-linux, but I do have a personal `aarch64-linux-gnu-gcc` symlink.)

**Comment by mmerah** [comment]
> Please check if this is still an issue. #27261 was merged two days ago. It significantly changed the build for non-windows targets, and it has ` --cross-compile-prefix=`. (arm64-linux builds for me on x64-linux, but I do have a personal `aarch64-linux-gnu-gcc` symlink.)

This is still an issue for me with the latest master. I still have this problem with the `cross-compile-prefix` while building.

**Comment by dg0yt** [comment]
Can you share your version of these lines from `buildtrees/openssl/arm64-linux-dbg/Makefile`:
~~~Makefile
CROSS_COMPILE=
CC=$(CROSS_COMPILE)/home/dg0yt/bin/aarch64-linux-gnu-gcc
CXX=$(CROSS_COMPILE)/home/dg0yt/bin/aarch64-linux-gnu-g++
~~~
and from `buildtrees/openssl/cmake-get-vars-arm64-linux-dbg.cmake.log`:
~~~
set(VCPKG_DETECTED_CMAKE_C_COMPILER "/home/dg0yt/bin/aarch64-linux-gnu-gcc")
...
set(VCPKG_DETECTED_CMAKE_CXX_COMPILER "/home/dg0yt/bin/aarch64-linux-gnu-g++")
~~~

**Comment by mmerah** [comment]
> Can you share your version of these lines from `buildtrees/openssl/arm64-linux-dbg/Makefile`:
> 
> ```makefile
> CROSS_COMPILE=
> CC=$(CROSS_COMPILE)/home/dg0yt/bin/aarch64-linux-gnu-gcc
> CXX=$(CROSS_COMPILE)/home/dg0yt/bin/aarch64-linux-gnu-g++
> ```
> 
> and from `buildtrees/openssl/cmake-get-vars-arm64-linux-dbg.cmake.log`:
> 
> ```
> set(VCPKG_DETECTED_CMAKE_C_COMPILER "/home/dg0yt/bin/aarch64-linux-gnu-gcc")
> ...
> set(VCPKG_DETECTED_CMAKE_CXX_COMPILER "/home/dg0yt/bin/aarch64-linux-gnu-g++")
> ```

buildtrees/openssl/arm64-linux-dbg/Makefile:
```
CROSS_COMPILE=aarch64-linux-gnu-
CC=$(CROSS_COMPILE)/usr/bin/aarch64-linux-gnu-gcc
CXX=$(CROSS_COMPILE)/usr/bin/aarch64-linux-gnu-g++
```

buildtrees/openssl/cmake-get-vars-arm64-linux-dbg.cmake.log:
```
set(VCPKG_DETECTED_CMAKE_C_COMPILER "/usr/bin/aarch64-linux-gnu-gcc")
set(VCPKG_DETECTED_CMAKE_CXX_COMPILER "/usr/bin/aarch64-linux-gnu-g++")
```

The CROSS_COMPILE variable should be empty but isn't

**Comment by dg0yt** [comment]
`--cross-compile-prefix` handling is fixed with #29966. Can you test this PR? For me, it results in `CROSS_COMPILE=` for arm64-linux.

**Comment by hushanjushi** [comment]
> `--cross-compile-prefix` handling is fixed with #29966. Can you test this PR? For me, it results in `CROSS_COMPILE=` for arm64-linux.

Works for me.

**Comment by mmerah** [comment]
> * docker build -t openssl-vcpkg-crosscompile:latest .

The `--cross-compile-prefix` behavior now works properly with this PR! Thanks!
